tokio-postgres/src/passfile/mod.rs

```diff
@@ -0,0 +1,192 @@
+
+fn field_matches(search_value: &[u8], file_value: &[u8]) -> bool {
+    if file_value == [b'*'] {
+        return true;
+    }
+    file_value == search_value
+}
+
+/// Removes trailing CR and/or LF from a string.
+///
+/// Intended to match libpq's pg_strip_crlf().
+fn strip_crlf(bb: &[u8]) -> &[u8] {
+    for (idx, b) in bb.iter().copied().enumerate().rev() {
+        if b != b'\n' && b != b'\r' {
+            return &bb[..=idx];
+        }
+    }
+    &bb[..0]
+}
+
+/// Searches passfile text for a match.
+///
+/// Intended to match libpq's behavior closely.
+/// If there is an IO error, returns None.
+async fn password_for_key<T>(key: &PassfileKey<'_>, buf: T) -> Option<Vec<u8>>
+where
+    T: tokio::io::AsyncBufReadExt + Unpin,
+{
+    let mut it = buf.split(b'\n');
+    // If we get an io error, just stop reading
+    while let Ok(Some(line)) = it.next_segment().await {
+        if line.starts_with(&[b'#']) {
+            continue;
+        }
+        let line = strip_crlf(&line);
+        if line.is_empty() {
+            continue;
+        }
+        if let Ok(entry) = PassfileEntry::new(line) {
+            if field_matches(key.hostname, &entry.hostname)
+                && field_matches(&key.port, &entry.port)
+                && field_matches(key.dbname, &entry.dbname)
+                && field_matches(key.user, &entry.user)
+            {
+                if entry.password.is_empty() {
+                    // To be consistent with libpq, in this case we need to
+                    // stop searching the password file, but not attempt to
+                    // use the empty password string.
+                    return None;
+                } else {
+                    return Some(entry.password);
+                }
+            }
+        }
+    }
+    None
+}
+
+/// Searches a passfile, returning the password from the first matching line.
+///
+/// Returns None if:
+///  - there is no match
+///  - the passfile doesn't exist
+///  - there is an error reading the passfile
+///  - `config` doesn't have a user set.
+///
+/// If `config` doesn't have a dbname set, matches the dbname field against the user name.
+///
+/// Intended to match libpq's behavior closely.
+///
+/// Unlike libpq, doesn't attempt to convert a `host` set to the default socket path into `"localhost"` for matching
+/// purposes. So passfiles using `localhost` to supply passwords for unix socket connections won't work.
+///
+/// Unlike libpq, doesn't enforce any rules on the passfile's permissions.
+pub(crate) async fn find_password(
+    passfile: &Path,
+    config: &Config,
+    host: &Host,
+    port: u16,
+) -> Option<Vec<u8>> {
+    // If we don't have a user, this connection can never work anyway
+    if let Some(user) = config.get_user() {
+        #[cfg(unix)]
+        {
+            let meta = match fs::metadata(&passfile).await {
+                Ok(meta) => meta,
+                Err(_) => return None,
+            };
+            if !meta.is_file() {
```

Comment:
Is this to check if the file is a symlink or something? Why would we only do that on Unix?

Reply:
I'm imitating libpq there (`passwordFromFile` in `fe-connect.c`). It doesn't reject symlinks. I'm not sure what it's for; maybe it's related to the following permissions check.